diff --git a/node_modules/next/dist/server/render.js b/node_modules/next/dist/server/render.js
index ed5eda3..6d4a57b 100644
--- a/node_modules/next/dist/server/render.js
+++ b/node_modules/next/dist/server/render.js
@@ -815,9 +815,14 @@ async function renderToHTMLImpl(req, res, pathname, query, renderOpts, extra) {
         // Always using react concurrent rendering mode with required react version 18.x
         const renderShell = async (EnhancedApp, EnhancedComponent)=>{
             const content = renderContent(EnhancedApp, EnhancedComponent);
-            return await (0, _nodewebstreamshelper.renderToInitialFizzStream)({
-                ReactDOMServer: _serveredge.default,
-                element: content
+            return new Promise((resolve, reject) => {
+                (0, _nodewebstreamshelper.renderToInitialStream)({
+                    ReactDOMServer: _serverbrowser.default,
+                    element: content,
+                    streamOptions: {
+                        onError: reject
+                    }
+                }).then(resolve, reject);
             });
         };
         const hasDocumentGetInitialProps = process.env.NEXT_RUNTIME !== 'edge' && !!Document.getInitialProps;
