diff --git a/node_modules/@codesandbox/sandpack-react/dist/esm/index.js b/node_modules/@codesandbox/sandpack-react/dist/esm/index.js
index 4acfc3a..d00e57d 100644
--- a/node_modules/@codesandbox/sandpack-react/dist/esm/index.js
+++ b/node_modules/@codesandbox/sandpack-react/dist/esm/index.js
@@ -566,17 +566,16 @@ var REACT_TEMPLATE = {
     },
     "/index.js": {
       code: `import React, { StrictMode } from "react";
-import ReactDOM from "react-dom";
+import { createRoot } from "react-dom/client";
 import "./styles.css";
 
 import App from "./App";
 
-const rootElement = document.getElementById("root");
-ReactDOM.render(
+const root = createRoot(document.getElementById("root"));
+root.render(
   <StrictMode>
     <App />
-  </StrictMode>,
-  rootElement
+  </StrictMode>
 );`
     },
     "/styles.css": {
@@ -611,8 +610,8 @@ h1 {
     }
   },
   dependencies: {
-    react: "^17.0.0",
-    "react-dom": "^17.0.0",
+    react: "^18.0.0",
+    "react-dom": "^18.0.0",
     "react-scripts": "^4.0.0"
   },
   entry: "/index.js",
